---
title: "Microbiome Workshop"
author: "Margaret Turner"
date: "3/4/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

See the original vignette [here](http://biocworkshops2019.bioconductor.org.s3-website-us-east-1.amazonaws.com/page/MicrobiomeWorkshop__MicrobiomeWorkshop/).

```{r}
#Load packages
library(MicrobiomeWorkshop)
library(curatedMetagenomicData)
library(HMP16SData)
library(HMP2Data)
library(phyloseq)
library(SummarizedExperiment)
library(MultiAssayExperiment)
library(dplyr)
library(ggplot2)
library(UpSetR)
library(ade4)
library(vegan)
library(MiRKAT)
```

Loading in data:

```{r}
#Load 16S data as a matrix, rows are Greengene IDs, columns are sample names:
data("momspi16S_mtx")

#Load the Greengenes taxonomy table as a matrix, rows are Greengene IDs, columns are taxonomic ranks:
data("momspi16S_tax")
# Check if Greengene IDs match between the 16S and taxonomy data
all.equal(rownames(momspi16S_mtx), rownames(momspi16S_tax)) # Should be TRUE

#Load the 16S sample annotation data as a matrix, rows are samples, columns are annotations:
data("momspi16S_samp")
# Check if sample names match between the 16S and sample data
all.equal(colnames(momspi16S_mtx), rownames(momspi16S_samp)) # Should be TRUE

#The MOMS-PI cytokine data can be loaded as a matrix, rownames are cytokine names, colnames are sample names:
data("momspiCyto_mtx")

#Load the cytokine sample annotation data as a matrix, rows are samples, columns are annotations:
data("momspiCyto_samp")
```



```{r}
#The momspi16S function assembles those matrices into a phyloseq object.
momspi16S_phyloseq <- momspi16S()
momspi16S_phyloseq

#The function momspiCytokines will make a SummarizedExperiment containing cytokine data
momspiCyto <- momspiCytokines()
momspiCyto
```

IBD Data

```{r}
#Load 16S data as a matrix, rows are SILVA IDs, columns are sample names:
data("IBD16S_mtx")

#Load the SILVA taxonomy table as a matrix, rows are SILVA IDs, columns are taxonomic ranks:
data("IBD16S_tax")

#Load the 16S sample annotation data as a matrix, rows are samples, columns are annotations:
data("IBD16S_samp")

#The IBD phyloseq object can be loaded as follows.
IBD <- IBD16S()
IBD
```

T2D

```{r}
#Load 16S data as a matrix, rows are Greengene IDs, columns are sample names:
data("T2D16S_mtx")

#Load the Greengenes taxonomy table as a matrix, rows are Greengene IDs, columns are taxonomic ranks:
data("T2D16S_tax")

#Load the 16S sample annotation data as a matrix, rows are samples, columns are annotations:
data("T2D16S_samp")

#The T2D phyloseq object can be loaded like so.
T2D <- T2D16S()
T2D
```

## COIA

### Pre-processing

```{r}
#Combine 16S and cytokines data
#order both sets by visit number within a subject
momspi16S_samp <- momspi16S_samp[
  with(momspi16S_samp, order(subject_id, sample_body_site, visit_number)),
] 

momspiCyto_samp <- momspiCyto_samp[
  with(momspiCyto_samp, order(subject_id, sample_body_site, visit_number)),
]

#select data collected at the same visit
combined_samp <- merge(momspi16S_samp, momspiCyto_samp, 
                       by = c("subject_id", "sample_body_site", 
                        "project_name", "study_full_name",
                        "subject_gender", "subject_race",
                        "visit_number"))

#select data from first visit only
combined_samp <- combined_samp[combined_samp$visit_number ==  1,]

table(combined_samp$sample_body_site)#all vaginal samples

#select 16S data for those samples
combined_16S_phyloseq <- subset_samples(momspi16S_phyloseq, file_name %in% combined_samp$file_name.x)

#get rif of otus that are not observed in any sample for this subset
combined_16S_phyloseq %<>%
    taxa_sums() %>%
    is_greater_than(0) %>%
    prune_taxa(combined_16S_phyloseq)
combined_16S_mtx <- otu_table(combined_16S_phyloseq)

#can get same data directly from matrices, but less convenient because phyloseq has taxonomy info
#combined_16S_mtx <- momspi16S_mtx[, colnames(momspi16S_mtx) %in% combined_samp$file_name.x]
#get rif of otus that are not observed in any sample for this subset
#combined_16S_mtx <- combined_16S_mtx[apply(combined_16S_mtx, 1, nnzero) >0, ]
combined_Cyto_mtx <- momspiCyto_mtx[, colnames(momspiCyto_mtx) %in% combined_samp$file_name.y ]
dim(combined_Cyto_mtx)

#Make sure that samples are in rows and variables (taxa and cytokines) are in columns.
combined_16S_mtx <- t(combined_16S_mtx)
combined_Cyto_mtx <- t(combined_Cyto_mtx)

#Taxa are converted to proportions.
combined_16S_mtx <- combined_16S_mtx/apply(combined_16S_mtx, 1, sum)
```


Let $X$ and $Y$ be the 16S and cytokines tables, respectively. The rows contain the same $n$ women's first visits. The columns contain $p_X$ taxa and $p_Y$ cytokines. The column weights are $Q_X$ and $Q_Y$, with row weights $D$. Then the PCA analysis of each table will be $(X, Q_X, D), (Y, Q_Y,D)$. The co-inertia axes will be generated from an eigendecomposiiton of $(YD^TX, Q_X, Q_Y)\implies Y^TDX=K\Lambda^{1/2}A^T$. Then we plot $F_X=XA, F_Y=YK$.

**Note: Why scale cytokines data but not taxa data? Plot histograms before this step.**

For taxa:

- Center 16S data to work with PCA on the covariance matrix $\Sigma_X=\text{Cov}X$.
- To normalize the magnitude, divide each value of $X$ by the total variance: $\sqrt{\mbox{tr}(\Sigma_X)}$.

Note that this second step is equivalent to dividing the matrix by $\sqrt{\sum_{k=1}^r \lambda_k}$, where $\lambda_k$ are the eigenvalues of $\Sigma_X$ and $r$ is the rank of $X$.

```{r}
taxa_mtx <- scale(combined_16S_mtx, center = TRUE, scale = FALSE)
#use fast trace computation formula: tr(A^B) = sum(A*B), where '*' operator refers to elementwise product
taxa_tr <- sum(taxa_mtx*taxa_mtx)/(dim(taxa_mtx)[1]-1)
taxa_mtx <- taxa_mtx/sqrt(taxa_tr)
taxa.pca <- ade4::dudi.pca(taxa_mtx, scannf=FALSE, nf =61,
                     center = FALSE, scale = FALSE)
```

For cytokines:

- Center and scale cytokines data to work with PCA on the correlation matrix $\Sigma_Y=\text{Corr}Y$.
- To normalize the magnitude, divide each value of $Y$ by the total variance: $\sqrt{\mbox{tr}(\Sigma_Y)}$.

```{r}
cyto_mtx <- scale(combined_Cyto_mtx, center = TRUE, scale = TRUE)
cyto_tr <- sum(cyto_mtx*cyto_mtx)/(dim(cyto_mtx)[1]-1)
cyto_mtx <- cyto_mtx/sqrt(cyto_tr)
cyto.pca <- ade4::dudi.pca(cyto_mtx, scannf=FALSE, nf =61,
                     center = FALSE, scale = FALSE)
```


### COIA

```{r}
#Co-inertia is available through R package ade4. It takes ade4 PCA objects and performs joint eigendecomposition.
coin <- ade4::coinertia(taxa.pca, cyto.pca, scannf = FALSE, nf = 2)
#RV coefficient â€“ measure of similarity between 16S and cytokines tables
RV <- coin$RV
RV
#> [1] 0.0396224
```

### Results

**Where are they getting this PlotCoinVars function??? I'll make one myself...**

```{r}
#p.vars <- PlotCoinVars(coin, tab1 = "taxa", tab2 = "cytokines", 
#             Labels1 = NULL, #colnames(combined_16S_mtx)
#             Labels2 = colnames(combined_Cyto_mtx),
#             label = TRUE,PtSize=2,LblSize=3,
#             hjust = 0, vjust = -1.5)
#p.vars
```


Interpretation of the plots is similar to examining PCA results.

Plot below provides variables projection on common co-inertia axes.

- Variables projected close to each other reflect similarilty within and across two data sets;
- Variables with larger values on each component (x- and y-axes) have more importance

We start by identifying important taxa (larger component values):

- visually from the plot, or
- by specifying x and y coordinates to pull all variables with coordinates larger (or smaller) than these values. 

Table 1 (taxa) values can be accessed `$co` object of coinertia output named `coin`.

```{r}
kableExtra::kable(head(coin$co),
      digits = 5)%>%
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover"), full_width = F)
```

Similarly, table 2 (here cytokines) variables loadings can be extracted from the $li object of coinertia output named coin.

```{r}
kableExtra::kable(head(coin$li),
      digits = 5)%>%
  kableExtra::kable_styling(bootstrap_options = c("striped", "hover"), full_width = F)
```


Samples with largest difference across two data sets. Samples with arrow lengths in 0.9 quantile are chosen.

```{r}
#Taxa with major differences across two sets
#large.difs <- rownames(Samp.coin$Dissimilarity[Samp.coin$Dissimilarity$Quantile >= 0.9, ])
```

Samples with largest differences across two data sets.

```{r}
#small.difs <- rownames(Samp.coin$Dissimilarity[Samp.coin$Dissimilarity$Quantile < 0.1, ])
```

